### ISSUE TITLE
Integration tests are not working

### ISSUE THREAD

**Issue by metamorphling** [ssannandeji/Zenject-2019#526]
Win10
Unity 2018.2.0f2
Zenject 7.1.0

Doing it by guide

> you can add an integration test for it by right clicking somewhere in the Project tab and then selecting Create -> Zenject -> Integration Test and then naming it SpaceShipTests.cs. This will create the following template code with everything you need to start writing your test:

Creating UntitledIntegrationTest.cs in Assets/Scripts leads to

> Assets/Scripts/UntitledIntegrationTest.cs(5,40): error CS0246: The type or namespace name `ZenjectIntegrationTestFixture' could not be found. Are you missing an assembly reference?

> Assets/Scripts/UntitledIntegrationTest.cs(3,19): error CS0234: The type or namespace name `TestTools' does not exist in the namespace `UnityEngine'. Are you missing an assembly reference?

Putting it under Assets/Scripts/Editor does not help(nor should it as far as I understand, since it's an integration test)
Copying TestFramework from Plugins into the Scripts helped to move a little forward, but opened yet another can of worms so I felt like this is not intended way in the first place.
Am I missing something?

Just for the record, had no problems with unit tests.

**Comment by svermeulen** [comment]
I think this is happening for you because in 2018 Unity added asmdef files and the new version of zenject includes these.  So what you can do is, in the directory where you have your integration tests, add an asmdef file and add Zenject-TestFramework as a dependency.  I'll add more documentation for this as well since you're right that the way the docs are written suggests that it should be easier

**Comment by metamorphling** [comment]
Thanks for the tip, I went into reading about asmdef files to update my knowledge, then tried proposed solution. I created an Assembly Definition, added Zenject-TestFramework in references. That allows integration test to find a ZenjectIntegrationTestFixture dependency, but it looses "UnityEngine.TestTools" namespace. Also tried:
1) Assembly Definition + references Zenject-TestFramework and Zenject-TestFramework-Editor(any platform)
2) Assembly Definition 1 + reference Zenject-TestFramework(any platform exluding editor), Assembly Definition 2 + reference Zenject-TestFramework-Editor(editor platform only)

Also, from what I understand about asmdef files they are meant to get rid of special folders, yet it still feels kind of clunky. I may be misunderstanding something, but having a step-by-step guide in documentation would be awesome.

**Comment by svermeulen** [comment]
Oh yeah, you also have to check 'Include Test Assemblies' in your custom asmdef.  Since you are using nunit.  You could also just delete all the asmdef files in the zenject folder and then execute 'enable playmode tests for all assemblies' from test runner.

I agree, they still feel pretty clunky, so maybe it was a bit premature to enable them in zenject in retrospect.  Since now everyone has to use them if they are also using Zenject

**Comment by svermeulen** [comment]
I'll add some information about it to the docs

**Comment by metamorphling** [comment]
Thanks, tried first test assemblies proposal and it worked! A small Unity 2018 step-by-step from me for anyone who gets lost on this one
1) Create -> Assembly Definition, for ex. under /Scripts/Tests/ (no need to have a special folder)
"References": Zenject-TestFramework
"Test Assemblies": Checked
2) Create -> Zenject -> Integration Test, should be under the same folder as asmdef file
3) Window -> General -> Test Runner -> PlayMode tab
Created test from step 2 should be visible under created asmdef from step 1, happy testing

**Comment by svermeulen** [comment]
No problem, glad you got it working.  I'll leave this open to add documentation for it

**Comment by metamorphling** [comment]
Celebration did not last long. As soon as I've tried to bring back my real world tests, things fell apart.
Adding
`using Assert = ModestTree.Assert;`
was enough to see that ModestTree is not getting found. When I've tried my own namespaces they were not getting found too. Shooting in the dark, I feel like folder with asmdef file gets pulled into an early build, like Assets/Plugins folder, and that makes it impossible to reference anything from Assets/Scripts

**Comment by svermeulen** [comment]
Yeah, maybe you should just delete all the asmdef files in Zenject and then "enable playmode tests" from the test runner?  Then it should work just like before.

When you change to use asmdef files I think you have to really commit to doing it everywhere

**Comment by svermeulen** [comment]
Or if you wanted to keep using asmdef files, you'd have to add a reference to the Zenject asmdef in addition to the test framework asmdef

**Comment by metamorphling** [comment]
> When you change to use asmdef files I think you have to really commit to doing it everywhere

Exactly, I have to wrap all my code into assemblies and be ready to commit myself to a new approach.
I believe Unity meant good with it, but I'm yet to see any significant merits to this. Maybe next time.

> Yeah, maybe you should just delete all the asmdef files in Zenject and then "enable playmode tests" from the test runner? Then it should work just like before.

This is good enough for my current situation

**Comment by adamsingle** [comment]
I can not seem to get this to compile. I've tried going down the path of setting up all the asmdef files and I get an error that the method subcontainercreatorbyinstaller does not contain a constructor that takes 2 arguments. If I delete all the asmdef files, I get compile errors that SceneTestFixture has no reference to the NUnit or UnityEngine.TestTools namespaces. 
Did I miss something in the docs? Something I haven't set up?

**Comment by adamsingle** [comment]
Hmm. I had to move both ZenjectIntegrationTestFixture and ZenjectSceneTestFixture from Zenject/OptionalExtras/TestFramework to Zenject/OptionalExtras/TestFramework/Editor to be next to ZenjectUnitTestFixture.cs (to remove those above mentioned errors). This has now allowed the other error I was seeing with the asmdef files:
'SubContainerCreatorByInstaller' does not contain a constructor that takes 2 arguments being thrown in SubContainerInstallerBindingFinalizer.cs on line 38

**Comment by adamsingle** [comment]
I feel I must have done some part of the installation incorrectly. No one else seems to have had issues with where files were located for example. Though I can't think what that could be. I installed Zenject from the asset store. I then wanted to unzip the automocking. So I've unzipped that directly into Zenject/OptionalExtras/TestFramework/Editor as mentioned in the Readme. This includes the contents of the Moq zip also. Is that correct?

**Comment by svermeulen** [comment]
What version of zenject / unity?

I just tried doing this and it worked:
- Installed zenject version 7.2.0 with Unity 2018.2.4f1
- Opened test runner
- Clicked PlayMode
- Clicked Create PlayMode Test Folder
- Inside the folder, click on Tests.amsdef
- Add Zenject-TestFramework to references
- Inside that folder, right click -> Create -> Zenject -> Integration Test
- Run playmode tests without issue

**Comment by adamsingle** [comment]
I think it's possible I had parts of two versions (7.1 and 7.2)

I just tried deleting Zenject completely
I checked that the import showed everything as new to be sure I hadn't missed anything.
Re-imported. Re-unzipped automock zip and moq lib. 
Added the references to test.asmdef

no errors. I'm not sure exactly what I did wrong, but I appreciate your patience. I look forward to writing some tests now!
